### PR TITLE
[CVP] Add `samesign` flag to `icmp`

### LINKED ISSUE #114820: [CVP] Infer samesign in CorrelatedValuePropagation
CVP performs signed -> unsigned comparison canonicalization, so we should also infer the samesign flag there. Relevant function is: https://github.com/llvm/llvm-project/blob/3268d51a5c81be705a97216bf061fab9bb195ebe/llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp#L286

We need a change similar to https://github.com/llvm/llvm-project/commit/095d49da76be09143582e07a807c86d3b4334dec for InstCombine.

### PR DESCRIPTION
**Pull request by rhanqtl**
Closes #114820

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 1]
```diff
@@ -288,24 +289,37 @@ static bool processICmp(ICmpInst *Cmp, LazyValueInfo *LVI) {
...
+      return false;
+
+    ++NumSICmps;
+    Cmp->setPredicate(UnsignedPred);
+    Changed = true;
+  }
+
+  auto Op0Known = CR1.toKnownBits(), Op1Known = CR2.toKnownBits();
+  if ((Op0Known.Zero.isNegative() && Op1Known.Zero.isNegative()) ||
+      (Op0Known.One.isNegative() && Op1Known.One.isNegative())) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:317]:
```suggestion
  if ((CR1.isAllNonNegative() && CR2.isAllNonNegative()) ||
      (CR1.isAllNegative() && CR2.isAllNegative())) {
```

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 2]
```diff
@@ -38,6 +38,7 @@
 #include "llvm/IR/Type.h"
 #include "llvm/IR/Value.h"
 #include "llvm/Support/Casting.h"
+#include "llvm/Support/KnownBits.h"
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:41]:
```suggestion
```

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 3]
```diff
@@ -288,24 +289,37 @@ static bool processICmp(ICmpInst *Cmp, LazyValueInfo *LVI) {
...
-          LVI->getConstantRangeAtUse(Cmp->getOperandUse(1),
-                                     /*UndefAllowed*/ true));
+  bool Changed = false;
 
-  if (UnsignedPred == ICmpInst::Predicate::BAD_ICMP_PREDICATE)
-    return false;
+  ConstantRange CR1 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(0),
+                                                 /*UndefAllowed*/ true),
+                CR2 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(1),
+                                                 /*UndefAllowed*/ true);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:300]:
```suggestion
  ConstantRange CR1 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(0),
                                                 /*UndefAllowed=*/true),
                CR2 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(1),
                                                 /*UndefAllowed=*/true);
```

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 4]
```diff
@@ -288,24 +288,36 @@ static bool processICmp(ICmpInst *Cmp, LazyValueInfo *LVI) {
...
-          LVI->getConstantRangeAtUse(Cmp->getOperandUse(1),
-                                     /*UndefAllowed*/ true));
+  bool Changed = false;
 
-  if (UnsignedPred == ICmpInst::Predicate::BAD_ICMP_PREDICATE)
-    return false;
+  ConstantRange CR1 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(0),
+                                                 /*UndefAllowed=*/ true),
+                CR2 = LVI->getConstantRangeAtUse(Cmp->getOperandUse(1),
+                                                 /*UndefAllowed=*/ true);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:299]: I don't think that UndefAllowed=true is correct for this optimization. We should set UndefAllowed=false.

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 5]
```diff
@@ -288,24 +288,36 @@ static bool processICmp(ICmpInst *Cmp, LazyValueInfo *LVI) {
...
+    if (UnsignedPred == ICmpInst::Predicate::BAD_ICMP_PREDICATE)
+      return false;
+
+    ++NumSICmps;
+    Cmp->setPredicate(UnsignedPred);
+    Changed = true;
+  }
+
+  if ((CR1.isAllNonNegative() && CR2.isAllNonNegative()) ||
+      (CR1.isAllNegative() && CR2.isAllNegative())) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:315]:
```suggestion
  if (ConstantRange::areInsensitiveToSignednessOfICmpPredicate(CR1, CR2)) {
```
I just forgot this one.

**Review by dtcxzyw — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 6]
```diff
@@ -288,24 +288,36 @@ static bool processICmp(ICmpInst *Cmp, LazyValueInfo *LVI) {
   if (!Cmp->getOperand(0)->getType()->isIntOrIntVectorTy())
     return false;
 
-  if (!Cmp->isSigned())
+  if (!(Cmp->isSigned() || (Cmp->isUnsigned() && !Cmp->hasSameSign())))
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:291]:
```suggestion
  if (!Cmp->isSigned() && (!Cmp->isUnsigned() || Cmp->hasSameSign())))
```
nit: Prefer pushing `!` into conditions.

**Comment by nikic** [comment]
It looks like you also need to update a clang test:
```
^[_bk;t=1731230406738^GFailed Tests (1):^M
^[_bk;t=1731230406738^G  Clang :: CodeGen/aarch64-pure-scalable-args.c^M
```

**Review by nikic — APPROVED** [review]
LGTM

**Review by antoniofrighetto — APPROVED** [review]

**Pull request merged**